src/source/Ice/ConnectionListener.c

```diff
@@ -116,17 +117,34 @@ STATUS connectionListenerAddConnection(PConnectionListener pConnectionListener,
 STATUS connectionListenerRemoveConnection(PConnectionListener pConnectionListener, PSocketConnection pSocketConnection)
 {
     STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+    const UINT64 longTimeout = 1 * HUNDREDS_OF_NANOS_IN_AN_HOUR;
 
     CHK(pConnectionListener != NULL && pSocketConnection != NULL, STATUS_NULL_ARG);
     CHK(!ATOMIC_LOAD_BOOL(&pConnectionListener->terminate), retStatus);
 
-    // mark socket as closed. Will be cleaned up by connectionListenerReceiveDataRoutine
+    /* mark socket as closed. Will be cleaned up by connectionListenerReceiveDataRoutine */
     CHK_STATUS(socketConnectionClosed(pSocketConnection));
 
     ATOMIC_STORE_BOOL(&pConnectionListener->connectionListChanged, TRUE);
 
+    MUTEX_LOCK(pConnectionListener->lock);
+    locked = TRUE;
+    /* make sure connectionListenerRemoveConnection return after connectionListenerReceiveDataRoutine has picked up
+     * the change. */
+    while(ATOMIC_LOAD_BOOL(&pConnectionListener->listenerRoutineStarted) &&
+          !ATOMIC_LOAD_BOOL(&pConnectionListener->terminate) &&
+          ATOMIC_LOAD_BOOL(&pConnectionListener->connectionListChanged)) {
+        /* use longTimeout because connectionListenerReceiveDataRoutine should wake this up soon enough */
+        CVAR_WAIT(pConnectionListener->removeConnectionComplete, pConnectionListener->lock, longTimeout);
```

Comment:
I don't understand why 1 hour? Something more sensible perhaps with a handling of a timeout error?

Reply:
the point is it should never timeout. the connectionListenerReceiveDataRoutine should wake it up in time.

Reply:
Perhaps, but if it does? What then? The worse could happen if the code "hangs". Perhaps you need to set something small and error out on a timeout.

Reply:
I kind of prefer it hangs so we can get to it right away instead of chasing red herring.

Reply:
reduced to 5 seconds

Reply:
au contraire, on a production device, chasing hangs is impossible. I think you are doing the right thing in reducing and erring out.